src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -49,25 +50,44 @@ private void refreshCacheItem(final MessageCacheItem cacheItem, Iterator<DataSou
 			return;
 		}
 
-		long timestampOld = cacheItem.getTimestamp();
 		DataSourceEnum dataSource = msgSourceQueueIter.next();
-		String localeOrig = dto.getLocale();
-		if (dataSource.equals(DataSourceEnum.VIP) && dto.getLocale().equals(ConstantsKeys.SOURCE)) {
-			dto.setLocale(ConstantsKeys.LATEST);
-		}
-		dataSource.createMessageOpt(dto).getComponentMessages(cacheItem);
-		long timestamp = cacheItem.getTimestamp();
-		if (timestampOld == timestamp) {
-			logger.debug(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED, dto.getComponent(), dto.getLocale(), dataSource.toString()));
-		}
-		dto.setLocale(localeOrig);
+		if (!proceed(dataSource)) {
+			refreshCacheItem(cacheItem, msgSourceQueueIter); // Try the next dataSource
+		} else {
+			long timestampOld = cacheItem.getTimestamp();
+			String localeOrig = dto.getLocale();
+			if (dataSource.equals(DataSourceEnum.VIP) && dto.getLocale().equals(ConstantsKeys.SOURCE)) {
+				dto.setLocale(ConstantsKeys.LATEST);
+			}
+			dataSource.createMessageOpt(dto).getComponentMessages(cacheItem);
+			long timestamp = cacheItem.getTimestamp();
+			if (timestampOld == timestamp) {
+				logger.debug(FormatUtils.format(ConstantsMsg.GET_MESSAGES_FAILED, dto.getComponent(), dto.getLocale(), dataSource.toString()));
+			}
+			dto.setLocale(localeOrig);
 
-		// If timestamp is 0, it means that cacheItem not yet in cache. So try the next data source.
-		if (timestamp == 0) {
-			// Try the next dataSource in the queue
-			refreshCacheItem(cacheItem, msgSourceQueueIter);
+			// If timestamp is 0, it means that cacheItem not yet in cache. So try the next data source.
+			if (timestamp == 0) {
+				// Try the next dataSource in the queue
+				refreshCacheItem(cacheItem, msgSourceQueueIter);
+			}
 		}
+	}
+
+	/**
+	 * @return 'true' for either of the following cases. Otherwise, false (locale not supported in data source).
+	 * <ul>
+	 * 	<li>the dataSource's set of supported locales is not in cache. If the list is not in cache, it should not block refreshCacheItem</li>
+	 * 	<li>the requested locale is found in the data source's cached set of supported locales.</li>
+	 * </ul>
+	 */
+	private boolean proceed(DataSourceEnum dataSource) {
+		ProductService ps = new ProductService(dto);
+		Set<Locale> supportedLocales = ps.getSupportedLocales(false, dataSource);
+		logger.debug("supported languages: [{}]", supportedLocales.toString());
```

Comment:
No need to call "toString()" method as formatting and string conversion is done by the Formatter.